helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -376,62 +614,158 @@ public Object removeRoutingTableChangeListener(
    * @return
    */
   public Collection<LiveInstance> getLiveInstances() {
-    return _routingTableRef.get().getLiveInstances();
+    String key = getRoutingTableKey();
+    return _routingTableRef.get(key).get().getLiveInstances();
+  }
+
+  public Collection<LiveInstance> getLiveInstances(PropertyType propertyType) {
+    String key = getRoutingTableKey(propertyType);
+    return _routingTableRef.get(key).get().getLiveInstances();
+  }
+
+  public Collection<LiveInstance> getLiveInstances(PropertyType propertyType, String type) {
+    String key = getRoutingTableKey(propertyType, type);
+    return _routingTableRef.get(key).get().getLiveInstances();
   }
 
   /**
    * Return all instance's config in this cluster.
    * @return
    */
   public Collection<InstanceConfig> getInstanceConfigs() {
-    return _routingTableRef.get().getInstanceConfigs();
+    String key = getRoutingTableKey();
+    return _routingTableRef.get(key).get().getInstanceConfigs();
+  }
+
+  public Collection<InstanceConfig> getInstanceConfigs(PropertyType propertyType) {
+    String key = getRoutingTableKey(propertyType);
+    return _routingTableRef.get(key).get().getInstanceConfigs();
+  }
+
+  public Collection<InstanceConfig> getInstanceConfigs(PropertyType propertyType, String type) {
+    String key = getRoutingTableKey(propertyType, type);
+    return _routingTableRef.get(key).get().getInstanceConfigs();
   }
 
   /**
    * Return names of all resources (shown in ExternalView) in this cluster.
    */
   public Collection<String> getResources() {
-    return _routingTableRef.get().getResources();
+    String key = getRoutingTableKey();
+    return _routingTableRef.get(key).get().getResources();
+  }
+
+  public Collection<String> getResources(PropertyType propertyType) {
+    String key = getRoutingTableKey(propertyType);
+    return _routingTableRef.get(key).get().getResources();
+  }
+
+  public Collection<String> getResources(PropertyType propertyType, String type) {
+    String key = getRoutingTableKey(propertyType, type);
+    return _routingTableRef.get(key).get().getResources();
+  }
+
+  /**
+   * Function that returns the key associated with the specific _routingTableRef
+   * @return
+   */
+  private String getRoutingTableKey() {
+    // Check whether there exist only one snapshot (_routingTableRef)
+    if (_routingTableRef.keySet().size() == 1) {
+      return _routingTableRef.keySet().iterator().next();
+    } else {
+      throw new HelixException("There is none or more than one RoutingTableSnapshot");
+    }
+  }
+
+  private String getRoutingTableKey(PropertyType propertyType) {
+    if (propertyType == PropertyType.CUSTOMIZEDVIEW) {
+      throw new HelixException("Specific type needs to be used for CUSTOMIZEDVIEW PropertyType");
+    } else {
+      return getRoutingTableKey(propertyType, "");
+    }
+  }
+
+  private String getRoutingTableKey(PropertyType propertyType, String aggregationType) {
```

Comment:
Just keep this one. For the other 2 usages, just merge the check logic inside here. Put default values in the caller if necessary.

Reply:
Done.